graylog2-server/src/main/java/org/graylog2/users/UserImpl.java

```diff
@@ -144,11 +153,53 @@ public boolean isReadOnly() {
         return Collections.emptyMap();
     }
 
+    @Override
+    public String getFirstName() {
+        final Object firstName = fields.get(FIRST_NAME);
+        return firstName != null ? String.valueOf(firstName) : null;
+    }
+
+    @Override
+    public String getLastName() {
+        final Object lastName = fields.get(LAST_NAME);
+        return lastName != null ? String.valueOf(lastName) : null;
+    }
+
+    /**
+     * @return The full user name formatted as "First Last".
+     * first_name and last_name fields were added in Graylog 4.1.
+     * So, they might not be present for users that existed before.
+     * Fall back to full_name when they are not present.
+     */
     @Override
     public String getFullName() {
+        final Object firstName = fields.get(FIRST_NAME);
+        final Object lastName = fields.get(LAST_NAME);
+        if (firstName != null && lastName != null ) {
+            return String.format(Locale.ENGLISH, "%s %s", firstName, lastName);
+        }
+
         return String.valueOf(fields.get(FULL_NAME));
     }
 
+    /**
+     * Set the user's fullName, firstName, and lastName. The fullName is composed with the {@link #FULL_NAME_FORMAT}
+     * @param firstName The user's first name.
+     * @param lastName The user's last name.
+     */
+    @Override
+    public void setFullName(final String firstName, final String lastName) {
+        Preconditions.checkArgument(StringUtils.isNotBlank(firstName), "A first_name value is required.");
+        Preconditions.checkArgument(StringUtils.isNotBlank(lastName), "A last_name value is required.");
```

Comment:
Somehow my comment got lost, so here it goes again:
If we want to stick to the pattern of calling a bunch of setters and then validating upon saving the model, we could add some validation by overriding `validate` in `PersistedServiceImpl` and add a validation checking if either first and last name or the full name have been set.
We would only need to do this if we split up the setters as I suggested though.

Reply:
I agree. We should at least perform some minimal validation when saving the user object. As this was coded before, it would have been possible to not set any name, which would result in inconsistent data. Now, I have added some name validation in `UserServiceImpl` service.

Reply:
The validation requires that a last name is provided, because that is required in all cases. It also ensures that the first and last name are provided as a pair if either is provided. We cannot require that both are present, because there are some cases where only the full name is available.

Reply:
@thll I thought and looked quite hard at the topic of removing the two-argument setter that sets the first, last and full name internally. But, I cannot see a solution that works without side-effects or too much possibility for inconsistency.

First, I cannot see a way that the `full_name` can be auto-calculated upon saving a user without causing side-effects. This works ok for the create case, but it gets really challenging on update. It makes sense to set a `full_name` if the value is not there. But if it is, we don't know if it's the old value that needs to be updated, or a deliberately set _new_ value that we should not overwrite. This can get really confusing, and I worry that it won't be clear to the caller. 

Another option was to require the caller to set their own first, last and full name (concatenated by the caller each time). This was the most clean getter/setter approach, but this felt like it provided too much opportunity for inconsistency. The fact that the UserImpl object internally wants to maintain the `full_name` seems like an internal concern.

This is further complicated that the UserImpl object now provides support for usages that only provide a `full_name` and also others that provide the first/last. In the end, is seems to me that the latest approach as coded is the best compromise solution, that also has the best chance for consistency. I agree, it's not ideal, but also the fact that the Graylog user can be mapped to external users that only have a full name or first/last seems to justify that. I can't see a way to make it cleaner.

Also, you will notice that the preconditions checks in this method are still there too. This enforces that when a user wants to provide a first/last, that both must indeed be non-null. I don't see a way around this, and it only seems to help.

If you have any suggestions for improvement, of if something here does not seem correct, I welcome the feedback.